File: backend/internal/domain/food_item.go
// Package domain defines the Go types that mirror the contract §4 data
// model. Each type is the wire shape the API emits and the row shape the
// store package reads — one type, two views, no DTO/Model split.
//
// The top-level field names are camelCase per contract §3.1 wire
// convention; the JSONB sub-document field names are snake_case verbatim
// (contract §3.1: "JSONB sub-document fields are returned verbatim with
// the underlying snake_case keys — they are the same shape the AI tool
// calls produce and consume").
//
// Depends on: standard library (time, encoding/json), google/uuid.
// Depended on by: internal/api (handlers serialise these to the wire),
// internal/store (decoders scan rows into these types).
// Why it exists: the contract defines one canonical shape per resource;
// duplicating that shape into per-package DTOs would be the source of the
// drift the contract document exists to prevent. One type, used at both
// boundaries, kept in lock-step with the contract by code review.
package domain

import (
	"time"

	"github.com/google/uuid"
)

// FoodItem mirrors contract §4.2 (the Canonical Food Object). The wire
// JSON keys are camelCase per §3.1; the embedded JSONB sub-documents
// (Category, UsageContext, etc.) keep their snake_case keys verbatim.
type FoodItem struct {
	ID             uuid.UUID      `json:"id"`
	UserID         uuid.UUID      `json:"userId"`
	CanonicalName  string         `json:"canonicalName"`
	DisplayName    string         `json:"displayName"`
	Quantity       *Quantity      `json:"quantity,omitempty"`
	Category       Category       `json:"category"`
	Attributes     map[string]any `json:"attributes"`
	Flexibility    Flexibility    `json:"flexibility"`
	UsageContext   UsageContext   `json:"usageContext"`
	InventoryState InventoryState `json:"inventoryState"`
	Sourcing       Sourcing       `json:"sourcing"`
	Metadata       Metadata       `json:"metadata"`
	CreatedAt      time.Time      `json:"createdAt"`
	UpdatedAt      time.Time      `json:"updatedAt"`
}

// Quantity is the JSONB { amount, unit } shape. Nullable on the row —
// callers see a nil pointer.
type Quantity struct {
	Amount float64 `json:"amount"`
	Unit   string  `json:"unit"`
}

// Category is the JSONB { primary, secondary? } shape. Primary is one of
// the nine-value enum (contract §4.2 "category_primary_known" CHECK).
type Category struct {
	Primary   string `json:"primary"`
	Secondary string `json:"secondary,omitempty"`
}

// Flexibility is the JSONB { substitution_allowed, acceptable_variants,
// strict } shape from contract §4.2.
type Flexibility struct {
	SubstitutionAllowed bool     `json:"substitution_allowed"`
	AcceptableVariants  []string `json:"acceptable_variants"`
	Strict              bool     `json:"strict"`
}

// UsageContext is the JSONB { role, required?, recipe_ids?, meal_plan_id?,
// shopping_list_id? } shape. Role is one of inventory|shopping|planned|
// ingredient — per ADR-0009 only inventory and shopping are written by
// v1 code, but the constraint admits all four.
type UsageContext struct {
	Role           string      `json:"role"`
	Required       *bool       `json:"required,omitempty"`
	RecipeIDs      []uuid.UUID `json:"recipe_ids,omitempty"`
	MealPlanID     *uuid.UUID  `json:"meal_plan_id,omitempty"`
	ShoppingListID *uuid.UUID  `json:"shopping_list_id,omitempty"`
}

// InventoryState is the JSONB { status, on_hand_amount, last_confirmed }
// shape. status is one of confirmed|likely|unknown|out.
type InventoryState struct {
	Status        string     `json:"status"`
	OnHandAmount  *float64   `json:"on_hand_amount"`
	LastConfirmed *time.Time `json:"last_confirmed"`
}

// Sourcing is the JSONB { store_affinity, bulk_allowed, generic_ok } shape.
type Sourcing struct {
	StoreAffinity *string `json:"store_affinity"`
	BulkAllowed   bool    `json:"bulk_allowed"`
	GenericOK     bool    `json:"generic_ok"`
}

// Metadata is the JSONB { created_by, confidence } shape. created_by is
// one of "ai" or "user".
type Metadata struct {
	CreatedBy  string  `json:"created_by"`
	Confidence float64 `json:"confidence"`
}

File: ios/SousChefKit/Sources/Domain/FoodItem.swift
//  FoodItem.swift
//
//  The Canonical Food Object — the single shape for inventory, shopping
//  items, planned ingredients, and recipe ingredients (contract §4.2,
//  behavior spec §2.1).
//
//  Depends on:     Foundation and every Food* sub-document type in this
//                  directory.
//  Depended on by: the API target's `ingredients()` method, the Plan tab's
//                  ingredients list, and (Phase E onward) the cookbook
//                  ingredient helper.
//  Why it exists:  this is the Swift mirror of the `food_items` table the
//                  data track owns (contract §4.2). Every wire field
//                  becomes a property at the precise contract type:
//                  UUID for ids, Date for timestamps, Decimal for amounts,
//                  enums for the closed sets. The decoder catches contract
//                  drift the moment a fixture stops round-tripping —
//                  rather than a silent UI render of garbage.
//
//                  Top-level wire keys are camelCase per contract §3.1
//                  (Go handlers translate the database snake_case column
//                  names at the wire boundary). The nested JSONB types
//                  keep their snake_case keys verbatim per the same
//                  paragraph.

import Foundation

/// FoodItem is the iOS-side mirror of the contract §4.2 `food_items` row.
/// The struct is `Identifiable` by `id` so SwiftUI ForEach can use it
/// directly, `Sendable` so it crosses actor boundaries (the APIClient
/// hands items to the main-actor view), and `Hashable` for cheap equality.
public struct FoodItem: Codable, Sendable, Hashable, Identifiable {
    /// id is the row's primary key — a UUIDv7 per contract §4.2.
    public let id: UUID

    /// userId is the owning Supabase user. The contract guarantees the
    /// caller only sees their own rows (RLS), so this is informational
    /// for the client.
    public let userId: UUID

    /// canonicalName is the lowercase singular generic form (e.g. "milk",
    /// "scallion"). The database enforces the lowercase invariant via a
    /// CHECK constraint; the iOS client trusts that and does not re-lower.
    public let canonicalName: String

    /// displayName is the human-readable form (e.g. "2% Milk", "Green
    /// Onions"). This is what the UI shows.
    public let displayName: String

    /// quantity is the optional amount-plus-unit pair. Null on rows the
    /// AI emitted without a measurement (e.g. "salt to taste").
    public let quantity: FoodQuantity?

    /// category is the 9-value primary category + optional secondary.
    public let category: FoodCategory

    /// attributes is the open JSONB map for ad-hoc annotations.
    public let attributes: FoodAttributes

    /// flexibility is the substitution-preferences sub-document.
    public let flexibility: FoodFlexibility

    /// usageContext carries the role enum (ADR-0009) and any cross-
    /// reference identifiers.
    public let usageContext: FoodUsageContext

    /// inventoryState carries the 4-value status enum plus the optional
    /// on-hand amount and last-confirmed timestamp.
    public let inventoryState: FoodInventoryState

    /// sourcing carries the store-affinity preferences.
    public let sourcing: FoodSourcing

    /// metadata carries the provenance enum and confidence score.
    public let metadata: FoodMetadata

    /// createdAt is the row's insertion time (RFC 3339 UTC on the wire,
    /// per contract §3.2).
    public let createdAt: Date

    /// updatedAt is the row's last modification time. The data track
    /// maintains it via a Postgres trigger.
    public let updatedAt: Date

    /// CodingKeys translates the wire's camelCase top-level field names
    /// to Swift's lowerCamelCase. The contract §3.1 says response bodies
    /// use camelCase except inside JSONB sub-documents, which keep their
    /// snake_case keys (handled by each nested type's own CodingKeys).
    enum CodingKeys: String, CodingKey {
        case id
        case userId
        case canonicalName
        case displayName
        case quantity
        case category
        case attributes
        case flexibility
        case usageContext
        case inventoryState
        case sourcing
        case metadata
        case createdAt
        case updatedAt
    }

    public init(
        id: UUID,
        userId: UUID,
        canonicalName: String,
        displayName: String,
        quantity: FoodQuantity?,
        category: FoodCategory,
        attributes: FoodAttributes,
        flexibility: FoodFlexibility,
        usageContext: FoodUsageContext,
        inventoryState: FoodInventoryState,
        sourcing: FoodSourcing,
        metadata: FoodMetadata,
        createdAt: Date,
        updatedAt: Date
    ) {
        self.id = id
        self.userId = userId
        self.canonicalName = canonicalName
        self.displayName = displayName
        self.quantity = quantity
        self.category = category
        self.attributes = attributes
        self.flexibility = flexibility
        self.usageContext = usageContext
        self.inventoryState = inventoryState
        self.sourcing = sourcing
        self.metadata = metadata
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}
